### PR TITLE
Add test coverage for UME

### PR DESCRIPTION
**Pull request by etherealsunshine**
## Description
Improves test coverage for the UME model class. Addresses #80 

**What's added**
* `test_malformed_smiles_behavior` - Documents how UME handles invalid SMILES strings
* `test_malformed_protein_behavior` - Tests UME's robustness with invalid protein sequences  

**Features**
* **Edge case documentation**: Tests 11 types of malformed SMILES (unclosed parentheses, invalid characters, empty strings, etc.)
* **Protein sequence validation**: Tests 10 types of invalid protein inputs (invalid amino acids, special characters, mixed case, etc.)

**Usage**

```bash
# Run the new robustness tests
pytest tests/lobster/model/test__ume.py::TestUME::test_malformed_smiles_behavior -v
pytest tests/lobster/model/test__ume.py::TestUME::test_malformed_protein_behavior -v

```

## Type of Change
- [x] Bug fix
- [ ] New feature
- [ ] Documentation update
- [ ] Performance improvement
- [x] Code refactoring

## Testing
- [x] Tests pass locally
- [ ] Added new tests for new functionality
- [x] Updated existing tests if needed

## Checklist
- [x] Code follows style guidelines
- [x] Self-review completed
- [ ] Documentation updated if needed
- [x] No breaking changes (or clearly documented)

### CODE REVIEW

**Comment by etherealsunshine** [comment]
Hi @ncfrey, I've added some tests addressing #80, considering some of the issues were addressed in existing PR's and added RDKit to the list of dependencies since it had probably been missed out. Let me know if this looks good!

File: pyproject.toml [hunk 1]
```diff
@@ -39,6 +39,7 @@ dependencies = [
     "litdata>=0.2.49",
     "seaborn>=0.13.2",
     "dgeb>=0.2.0",
+    "rdkit>=2025.3.3",
```
Comment (ncfrey) [inline pyproject.toml:42]: thanks! this is actually included in the confusingly named `mgm` (an internally used name) optional dependencies set: https://github.com/prescient-design/lobster/blob/main/pyproject.toml#L64

so this can be deleted.

File: tests/lobster/model/test__ume.py [hunk 2]
```diff
@@ -577,3 +577,156 @@ def test_load_checkpoint_disable_flash_attn_cpu_inference(self):
             # The important part is that the model loaded successfully and
             # use_flash_attn attribute is correctly set
             print(e)
+
+    def test_malformed_smiles_behavior(self):
```
Comment (ncfrey) [inline tests/lobster/model/test__ume.py:581]: these malformed input tests are great, thanks for adding them!

File: tests/lobster/model/test__ume.py [hunk 3]
```diff
@@ -577,3 +577,156 @@ def test_load_checkpoint_disable_flash_attn_cpu_inference(self):
...
+            if result["outcome"] == "accepted":
+                print(f"{case_type:20}: accepted (norm: {result['embedding_norm']:.3f})")
+            else:
+                print(f"{case_type:20}:  rejected ({result['error_type']})")
+
+        # Only fail if valid protein is rejected
+        if results["valid_protein"]["outcome"] != "accepted":
+            pytest.fail("Rejected a valid protein sequence")
+
+    def test_memory_usage_scaling(self):
```
Comment (ncfrey) [inline tests/lobster/model/test__ume.py:667]: this looks like it might be more performance profiling than unit testing? 

this could be added as a callback like https://github.com/prescient-design/lobster/blob/main/src/lobster/callbacks/_tokens_per_second_callback.py

or a function for profiling

File: tests/lobster/model/test__ume.py [hunk 4]
```diff
@@ -577,3 +577,156 @@ def test_load_checkpoint_disable_flash_attn_cpu_inference(self):
...
+
+            except Exception as e:
+                memory_data[batch_size] = {
+                    "status": "failed",
+                    "error_type": type(e).__name__,
+                    "error_msg": str(e)[:100],
+                }
+                break  # Stop at first failure
+
+        print("\n=== UME Memory Usage Scaling ===")
```
Comment (karinazad) [inline tests/lobster/model/test__ume.py:718]: it's good to avoid print statements in tests

File: tests/lobster/model/test__ume.py (hunk 3 above)
Comment (karinazad) [reply tests/lobster/model/test__ume.py:667]: agreed with Nathan, the logic seems a little overcomplicated for a unit test

Comment (etherealsunshine) [reply tests/lobster/model/test__ume.py:667]: got it, i'll refactor it then

File: tests/lobster/model/test__ume.py [hunk 5]
```diff
@@ -577,3 +579,83 @@ def test_load_checkpoint_disable_flash_attn_cpu_inference(self):
             # The important part is that the model loaded successfully and
             # use_flash_attn attribute is correctly set
             print(e)
+
+    @pytest.mark.parametrize(
+        "smiles,type_of_sequence",
```
Comment (karinazad) [inline tests/lobster/model/test__ume.py:584]: instead of having `type_of_sequence` you could use `pytest.param.id` which is specifically meant to add IDs for the test case

so somethine like
`pytest.param(..., id = "unclosed_parenthesis")`

File: tests/lobster/model/test__ume.py [hunk 6]
```diff
@@ -577,3 +579,83 @@ def test_load_checkpoint_disable_flash_attn_cpu_inference(self):
...
+            embeddings = ume.embed_sequences([smiles], "SMILES")
+            embedding_norm = torch.norm(embeddings).item()
+            # Just validate it's a proper tensor if it succeeds
+            assert isinstance(embeddings, torch.Tensor)
+            assert embeddings.shape == (1, ume.embedding_dim)
+            assert not torch.isnan(embeddings).any()
+            assert not torch.isinf(embeddings).any()
+            print(f"{type_of_sequence}: accepted (norm: {embedding_norm:.3f})")
+
+        except Exception as e:
```
Comment (karinazad) [inline tests/lobster/model/test__ume.py:612]: instead of using Exception inside a test, pls use `with pytest.raises(...)` for expected failures. Cases where we don't except failure don't need to be handled

File: tests/lobster/callbacks/test__memory_usage_callback.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
+import torch
+from typing import Any
+
+
+class MemoryUsageCallback:
```
Comment (karinazad) [inline tests/lobster/callbacks/test__memory_usage_callback.py:5]: I think Nathan meant implementing memory profiling for UME as a lightning callback in `src` not in `tests`. This could be a different MR though as it's worth comparing other pytorch or lightning memory profilers

Comment (karinazad) [reply tests/lobster/callbacks/test__memory_usage_callback.py:5]: we can remove this for now anf from the test below

**Review by ncfrey — APPROVED** [review]

**Comment by ncfrey** [comment]
@etherealsunshine approved and merged!
